runtime/src/main/java/org/corfudb/runtime/collections/UFOTxnBase.java

```diff
@@ -0,0 +1,369 @@
+            return this;
+        }
+
+        this.txnContext = txnContext.merge(table, key,
+                this::fixUpMetadata, new CorfuRecord<>(value, metadata));
+        return this;
+    }
+
+    /**
+     * Core logic for handling metadata modifications on mutations
+     * @param oldRecord - The previous fetched record that exists in the table.
+     * @param deltaUpdate - New record that will be inserted/updated.
+     * @param <K> - type of the key
+     * @param <V> - type of the value or payload
+     * @param <M> - type of the metadata
+     * @return the merged record that will be inserted into the table.
+     */
+    private <K extends Message, V extends Message, M extends Message>
+    CorfuRecord<V, M> fixUpMetadata(CorfuRecord<V, M> oldRecord, CorfuRecord<V, M> deltaUpdate) {
+        M deltaMetadata = deltaUpdate.getMetadata();
+            final Message.Builder builder = deltaMetadata.toBuilder();
+            long currentTime = System.currentTimeMillis();
+            for (Descriptors.FieldDescriptor fieldDescriptor : deltaMetadata.getDescriptorForType().getFields()) {
+                switch (fieldDescriptor.getName()) {
+                    case "revision":
+                        if (oldRecord == null || oldRecord.getMetadata() == null) {
+                            builder.setField(fieldDescriptor, 0L);
+                        } else {
+                            Long prevRevision = (Long) oldRecord.getMetadata().getField(fieldDescriptor);
+                            Long givenRevision = (Long)deltaMetadata.getField(fieldDescriptor);
+                            if ((givenRevision > 0 && // Validate revision only if set
+                                    prevRevision.longValue() == givenRevision.longValue())
+                            || givenRevision == 0) { // Do not validate revision if field isn't set
+                                builder.setField(fieldDescriptor, prevRevision + 1);
+                            } else {
+                                throw new StaleRevisionUpdateException(prevRevision, givenRevision);
+                            }
+                        }
+                        break;
+                    case "create_time":
+                        if (oldRecord == null) {
+                            builder.setField(fieldDescriptor, currentTime);
+                        }
+                        break;
+                    case "last_modified_time":
+                        builder.setField(fieldDescriptor, currentTime);
+                        break;
+                    default: // By default just merge any field not explicitly set with prev value
+                        if (oldRecord != null
+                                && oldRecord.getMetadata() != null
+                                && oldRecord.getMetadata().hasField(fieldDescriptor)
+                                && !deltaMetadata.hasField(fieldDescriptor)) {
```

Comment:
This fieldDescriptor is from deltaMetadata in the for loop, how could `deltaMetadata.hasField(fieldDescriptor)` be false? I think we need to loop over oldRecord's fieldDescriptor file descriptor as well for merge, right?

Reply:
hasField() will fire only if it has been set, different from whether it has the descriptor right?